modules/evm/src/lib.rs

```diff
@@ -467,21 +570,80 @@ impl<T: Trait> Module<T> {
 			AccountStorages::insert(address, index, value);
 		}
 
-		<Accounts<T>>::mutate(&address, |maybe_account_info| {
+		<Accounts<T>>::mutate(&address, |maybe_account_info| -> Result<(), ExitError> {
 			if let Some(AccountInfo {
 				contract_info: Some(contract_info),
+				storage_usage,
 				..
 			}) = maybe_account_info.as_mut()
 			{
 				match storage_change {
-					StorageChange::Added => contract_info.storage_count = contract_info.storage_count.saturating_add(1),
+					StorageChange::Added => {
+						contract_info.storage_count = contract_info.storage_count.saturating_add(1);
+					}
 					StorageChange::Removed => {
-						contract_info.storage_count = contract_info.storage_count.saturating_sub(1)
+						contract_info.storage_count = contract_info.storage_count.saturating_sub(1);
 					}
 					_ => (),
 				}
+
+				let additional_storage = Self::additional_storage(address);
+				*storage_usage = additional_storage;
+
+				if additional_storage != pre_additional_storage {
+					// get maintainer quota and pay for the additional_storage
+					let maintainer = T::AddressMapping::to_evm_address(&contract_info.maintainer)
+						.ok_or_else(|| ExitError::Other("storage quota not enough".into()))?;
+					<Accounts<T>>::mutate(&maintainer, |maybe_maintainer_account_info| -> Result<(), ExitError> {
+						if let Some(AccountInfo {
+							storage_quota: maintainer_storage_quota,
+							storage_usage: maintainer_storage_usage,
+							..
+						}) = maybe_maintainer_account_info.as_mut()
+						{
+							if additional_storage > pre_additional_storage {
```

Comment:
one way to avoid expect is do `checked_sub` and if it failed, we know `pre_additional_storage > additional_storage `